lib/phobos/actions/process_message.rb

```diff
@@ -64,6 +62,21 @@ def process_message(payload)
           end
         end
       end
+
+      def wait_for_next_retry(interval)
+        MAX_SLEEP_INTERVAL = 3
+        raise Phobos::AbortError if @listener.should_stop?
```

Comment:
This snippet:
```
  raise Phobos::AbortError if @listener.should_stop?
  @listener.send_heartbeat_if_necessary
```

Let us name it "raise or send heartbeat", is repeated 3 times in this block. It feels hacky. Something simpler is necessary.

Also, I do not like mutating variables that were given as input to a function, let's create a copy of it instead. 

How about

```ruby
  def snooze(interval:)
    current_interval = interval

    while current_interval > MAX_SLEEP_INTERVAL
      sleep_and_heartbeat(MAX_SLEEP_INTERVAL)
      current_interval -= MAX_SLEEP_INTERVAL
    end

    sleep_and_heartbeat(current_interval)
  end

  private

  def sleep_and_heartbeat(interval)
    return if interval <= 0

    sleep interval

    raise Phobos::AbortError if @listener.should_stop?
    @listener.send_heartbeat_if_necessary
  end
```

Reply:
`MAX_SLEEP_INTERVAL` can probably be set lower than 3, maybe even as low as `0.1`. One could think it should be configurable but I think too much configuration options can be bad especially when it is hard to explain what the option actually does.

I am not sure if there are performance implications of running 30 sleeps of 0.1 instead of 1 sleep of 3, but the lines of the block we are invoking is guarded by if statements so I am not worried about the block statements as such. Might want to investigate this.

Reply:
Hi @klippx, I'm looking at picking up where Ziru left off on this PR, curious to hear why it would be desirable to set the `MAX_SLEEP_INTERVAL` as low as `0.1`, is that just to accommodate users of Phobos who have a very short `session_timeout` configured?

Your other comments make sense, I'll work on addressing them and submit another PR.